### ISSUE TITLE
Ruby: Unit Rack 3 compatibility

### ISSUE THREAD

**Issue by travisbell** [nginx/unit#755]
Hey guys, I know @hongzhidao is aware of this (it was discussed on #562 [here](https://github.com/nginx/unit/issues/562#issuecomment-1225919235)) but I just thought I'd drop a new issue to mention that Rack 3 has now been released and it is not compatible with Unit.

It would be great to see Unit work with v3.

Thanks!

**Comment by tippexs** [comment]
@travisbell thanks for this. Do you have any example app you want to share? 
If not we can simply adapt the one we have with the use of `gem 'rack', '~> 3.0'`

**Comment by travisbell** [comment]
Hey @tippexs, indeed simply requiring 3.0 will prevent Unit from starting, so that's all you need to do.

**Comment by tippexs** [comment]
Link to the interim fix https://github.com/nginx/unit/issues/562#issuecomment-1225919235

**Comment by ac000** [comment]
So the [patch](https://github.com/nginx/unit/issues/562#issuecomment-1225919235) from @hongzhidao does indeed make things work with both rack < 3.0 **and** rack 3.0

Here's basically the same patch that applies cleanly with git-apply(1) to current unit.

```diff
diff --git a/src/ruby/nxt_ruby.c b/src/ruby/nxt_ruby.c
index f316d8a5..0c598470 100644
--- a/src/ruby/nxt_ruby.c
+++ b/src/ruby/nxt_ruby.c
@@ -480,18 +480,25 @@ nxt_ruby_rack_init(nxt_ruby_rack_init_t *rack_init)
 
     rackup = rb_protect(nxt_ruby_rack_parse_script,
                         (VALUE) (uintptr_t) rack_init, &state);
-    if (nxt_slow_path(TYPE(rackup) != T_ARRAY || state != 0)) {
+
+    if (nxt_slow_path(state != 0)) {
         nxt_ruby_exception_log(NULL, NXT_LOG_ALERT,
                                "Failed to parse rack script");
         return Qnil;
     }
 
-    if (nxt_slow_path(RARRAY_LEN(rackup) < 1)) {
-        nxt_alert(rack_init->task, "Ruby: Invalid rack config file");
-        return Qnil;
-    }
+    if (TYPE(rackup) == T_ARRAY) {
+
+        if (nxt_slow_path(RARRAY_LEN(rackup) < 1)) {
+            nxt_alert(rack_init->task, "Ruby: Invalid rack config file");
+            return Qnil;
+        }
+
+        return RARRAY_PTR(rackup)[0];
 
-    return RARRAY_PTR(rackup)[0];
+    } else {
+        return rackup;
+    }
 }
 
 ```

**Comment by ac000** [comment]
This has been fixed by dc9f592d6e7123d57924146dcbf1be80366bc98b

Thanks everyone!
